File: worker/crypto.py
"""
Decrypt mailbox secrets encrypted by the web app (web/lib/crypto.ts).

Same scheme both sides: AES-256-GCM, format
  "v1.<base64 iv>.<base64 ciphertext>.<base64 authTag>"
Key comes from MAILBOX_ENC_KEY (base64, 32 bytes) — identical in web + worker.
"""
import os
import base64
from cryptography.hazmat.primitives.ciphers.aead import AESGCM


def _key() -> bytes:
    k = os.environ.get("MAILBOX_ENC_KEY")
    if not k:
        raise RuntimeError("MAILBOX_ENC_KEY is not set")
    raw = base64.b64decode(k)
    if len(raw) != 32:
        raise RuntimeError("MAILBOX_ENC_KEY must be 32 bytes (base64-encoded)")
    return raw


def decrypt_secret(blob: str) -> str:
    version, iv_b, ct_b, tag_b = blob.split(".")
    if version != "v1":
        raise ValueError("unknown secret format")
    iv = base64.b64decode(iv_b)
    ct = base64.b64decode(ct_b)
    tag = base64.b64decode(tag_b)
    # Python's AESGCM expects ciphertext || tag concatenated.
    return AESGCM(_key()).decrypt(iv, ct + tag, None).decode("utf-8")

File: web/lib/crypto.ts
import crypto from "crypto";

// AES-256-GCM secret encryption. SERVER-ONLY (uses MAILBOX_ENC_KEY).
// Format: "v1.<base64 iv>.<base64 ciphertext>.<base64 authTag>"
// The Python worker decrypts the same format (worker/crypto.py).
//
// Generate the key once:  openssl rand -base64 32
// and set MAILBOX_ENC_KEY in web/.env.local AND worker/.env (same value).

function key(): Buffer {
  const k = process.env.MAILBOX_ENC_KEY;
  if (!k) throw new Error("MAILBOX_ENC_KEY is not set");
  const buf = Buffer.from(k, "base64");
  if (buf.length !== 32) throw new Error("MAILBOX_ENC_KEY must be 32 bytes (base64-encoded)");
  return buf;
}

export function encryptSecret(plain: string): string {
  const iv = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv("aes-256-gcm", key(), iv);
  const ct = Buffer.concat([cipher.update(plain, "utf8"), cipher.final()]);
  const tag = cipher.getAuthTag();
  return `v1.${iv.toString("base64")}.${ct.toString("base64")}.${tag.toString("base64")}`;
}

export function decryptSecret(blob: string): string {
  const [v, ivB, ctB, tagB] = blob.split(".");
  if (v !== "v1") throw new Error("unknown secret format");
  const decipher = crypto.createDecipheriv("aes-256-gcm", key(), Buffer.from(ivB, "base64"));
  decipher.setAuthTag(Buffer.from(tagB, "base64"));
  return Buffer.concat([decipher.update(Buffer.from(ctB, "base64")), decipher.final()]).toString("utf8");
}
